doc/design/volumes/nfsvolumestores.md

```diff
@@ -0,0 +1,106 @@
+
+The advantage to using the interface is the storage layer maintains consistency of the volumes regardless of the storage backend used.  For instance it checks all containers during `volume destroy` to see if the named volume is still referenced by another container (whether the container is powered `on` or `off`).
+
+[For reference](https://github.com/vmware/vic/blob/master/lib/portlayer/storage/volume.go#L36)
+```
+ 35 // VolumeStorer is an interface to create, remove, enumerate, and get Volumes.
+ 36 type VolumeStorer interface {
+ 37 »···// Creates a volume on the given volume store, of the given size, with the given metadata.
+ 38 »···VolumeCreate(op trace.Operation, ID string, store *url.URL, capacityKB uint64, info map[string][]byte) (*Volume, error)
+ 39
+ 40 »···// Destroys a volume 
+ 41 »···VolumeDestroy(op trace.Operation, vol *Volume) error
+ 42
+ 43 »···// Lists all volumes 
+ 44 »···VolumesList(op trace.Operation) ([]*Volume, error)
+  ...
+ 48 }   
+```
+
+When we create the NFS `VolumeStore`, we'll store the NFS target parameters (`host` + `path`) in the implementation's struct.  This is the only information we'll need to mount the NFS target on the container.
+
+```
+type NFSv3VolumeStore struct {
+ target *url.URL
+}
+```
+
+_The implementation is still being worked on.  The open question is whether the VCH appliance will mount the target to manipulate the NFS target, or use an NFS client implementation in userspace instead.  I'd much (*MUCH*) rather do the latter.  The `linux` VFS implementation throws `sync` errors when mounts are unavailable.  And we don't want to bring down the appliance because of a network hiccup.  NFS is a simple protocol and there is a public pkg which implements most of it.  Adding the few primitives we need shouldn't be that difficult, but more evaluation of the work required is needed_
+
+#### VolumeCreate
+In the vsphere model, a volume is a `.vmdk` backed block device.  Creation of a volume entails attaching a new disk to the VCH, preparing it with a filesystem, and detaching it.  The resulting `.vmdk` lives in its own folder in the volume store directory (specified during install w/ `vic-machine`).  We're going to follow the same model except there is nothing to prepare.  Each volume will be a directory (which the container client will mount directly) and live at the top of the volume store directory (which we will prepare during install).
+
+Some psuedo code.
+```
+func VolumeCreate() {
+// volPath := vicVolumePath(nameOfVolume)
+// mkdir volPath
+// return volPath
+}
+```
+#### VolumeDestroy
+Likewise destroying the volume is simply removing the volume's top level directory.
+```
+func VolumeDestroy() {
+// volPath := vicVolumePath(nameOfVolume)
+// rm -rf volPath
```

Comment:
It's psuedo code, but the client descends the dirents normally.  We'll supply a call to do that in our client.